### ISSUE TITLE
Network topology awareness

### ISSUE THREAD

**Issue by phillebaba** [spegel-org/spegel#669]
### Describe the problem to be solved

When doing lookups in the DHT the first peer found will be returned. The query does not take into consideration the network topology of the Kuberentes cluster. Most cloud providers have multiple zones per region. Network traffic within the same zone is usually faster and cheaper compared to cross zone traffic. Spegel should prioritize peers in the same zone over other peers when possible.

### Proposed solution to the problem

There has been some interesting research already done on the topic about partitioning.

https://research.protocol.ai/publications/enriching-kademlia-by-partitioning/monteiro2022.pdf

We need to decide on the best solution to partition the DHT to avoid added latency for lookups.

**Comment by craig-seeman** [comment]
I'm curious what your thoughts are on an (admittedly rather naive) but simple approach to topology. I could be misinterpreting the paper linked, but it almost sounds like this sort of feature currently does not exist in the libraries and would have to be implemented? I have not looked too deeply into the code that translate sha key on the p2p side to how we interface with containerd, so this may not even work though.

If this is the case however, the approach I was wondering about would be to essentially just leverage a majority of the existing code and simply (if some sort of spegel topology setting is enabled) advertise an additional image with a single kubernetes label added in the p2p key. This would obviously double the keyspace; but given the kubernetes best practices etc - this count shouldn't get too out of hand.

For example:
Image: 983487d9c4b7451b0e7d282114470d3a0ad50dc5e554971a4d1cda04acde670b

What could happen is spegel advertises out [983487d9c4b7451b0e7d282114470d3a0ad50dc5e554971a4d1cda04acde670b] as it currently does, but also advertise out a key of 
sha256("983487d9c4b7451b0e7d282114470d3a0ad50dc5e554971a4d1cda04acde670b" + "topology.kubernetes.io/zone: \"us-east-1c\"") [589aa1323f0a4834bcbfe3a50a157c4fdded821c79873e2a21e08d3e36654f42]


During an image pull - if the topology setting is enabled, the first lookup would be to the topology (in this case looking for hash 589aa1323f0a4834bcbfe3a50a157c4fdded821c79873e2a21e08d3e36654f42) if none is found (none on the local zone) a possible additional setting could be in place that would determine what to do next (try a cluster-wide lookup or pull from the repository directly instead) and would do a query for 983487d9c4b7451b0e7d282114470d3a0ad50dc5e554971a4d1cda04acde670b.

**Comment by phillebaba** [comment]
I do not think you are wrong with your analysis. The paper describes a situation that is a lot more complicated than anything that we are really facing. Spegel is aimed to run in private clusters so it does not have the same internet scale challenges. We only have to deal with a fixed set of regions, for most this would be 3 availability zones. For others it may be more if they are running clusters cross regions, but I find that this is a lot more uncommon.

While it would increase the advertised keys I do think the benefits from prioritizing peers in the same zone outweighs any cost. Either way we can easily run the benchmarks and verify that this is the case.

I have been looking around and have not found a good way to pass the topology key into the running pods. I have found a couple of issues in the Kubernetes repository but most people suggest using external projects to solve this. Are you aware of an easy way to do this?

I am open to making this an opt in feature initially, to start evaluating the cost of the increased advertised keys. Down the road if we come up with a smarter solution, we can always make changes.

**Comment by phillebaba** [comment]
Stumbled upon some more good research on this topic which seems to be linked to the original paper I linked.

https://asc.di.fct.unl.pt/~jleitao/thesis/MonteiroMSc.pdf

**Comment by craig-seeman** [comment]
Hey @phillebaba I was having a think the past few days and keep thinking about #670 wondering if there is some common approach that could solve the two issues. Unless my reading of the libp2p docs (and possible chatgpt hallucinations) are incorrect, I think we may be able to utilize the k/v "store" built into dht/kdht to store host information such as AZ opposed to just the cid/sha? 

https://github.com/libp2p/go-libp2p/blob/2a580cf7c8f53c66806100036c0d08bef0f343e0/core/routing/routing.go#L60

From what I can gather, though, there is no current way to call findproviders and include a specific host/peer k/v, so we'd have to query and get a list of multiple hosts and sort through their k/vs which would include the AZ until we find a suitable match.

What I'm wondering here too is with this approach, since we are able to see the number of hosts/providers of a specific key/image sha, if there are less than, say 3 provider hosts, we could do a secondary query to a host directly which would somehow publish a metric of it's current bandwidth utilization (I'm not sure if this would work with the k/v updating as those are not done in a rapid manner in which bandwidth utilization would be changing)

**Comment by vkoshkarovroku** [comment]
@craig-seeman We can't use the downward API here since the required topology label isn't available at the Pod level - it exists only on the Node.

@phillebaba I've also explored passing zone information directly to Pods. While certain policy engines can provide this capability, they're often impractical when required as a dependency for a single component.

Currently, the only way to achieve this without external tools is to statically define zones and deploy a DaemonSet per zone using node selectors. However, this introduces a challenge: your discovery mechanism must be capable of locating all Pods across the defined zones. I'm new to Spegel and don't know the implementation details yet.

**Comment by fedecz** [comment]
@vkoshkarovroku , if the spegel pod needs to know the AZ is in, it could query the k8s api to get its pod information, get the node the pod is running on, and then a new call to the k8s api to get the details (and AZ) of the node.
These calls would need to be done only once, at spegel pod startup.

Once the spegel pod knows which AZ is running on, it could prioritise it when querying for image hashes (no idea how this would be done haha)

**Comment by phillebaba** [comment]
Forwarding the nodes zone to Spegel is probably the easiest part, even without being able to use the downward API to pass the node zone to the pod. It might be  a bit more dirty but it can be solved.

No matter what solution we choose the biggest challenge will be to build  a solution to prioritize peers within the same zone. I am currently exploring the soft partitioning strategy suggested in the white paper I linked to weight Kademlia towards peers within the same zone. It is a bit tricky to figure out how to get down to this low level to mask the bits of the IDs in libp2p.

There are other possible solutions out there, but they would have to be proven to be scalable. A possible experimental solution while we wait for something better would be to add a metadata API that we could query while resolving peers. If the peer is in the same AZ we would immediately attempt to forward the request, but if it isn't the peer would be put in the end of a queue which would only be attempted when the peer limit is reached. The challenge is that in a large cluster there is no guarantee that any peers in the same AZ would be found given that there are hundreds of providers of a given content.

Either way any topology aware feature would start by being experimental to allow me to make fixes based on feedback from users.
